File: src/main/java/Decorator.java
interface IComponent {
    public void operate();
}

class ConcreteComponent implements IComponent {
    @Override
    public void operate() {
        System.out.println("operate");
    }
}

abstract class Decorator implements IComponent {
    private Decorator decorator;

    Decorator(Decorator decorator) {
        this.decorator = decorator;
    }

    @Override
    public void operate() {
        decorator.operate();
    }
}

class ComponentDecorator extends Decorator {
    public ComponentDecorator(Decorator decorator) {
        super(decorator);
    }

    public void decoratedOperation() {
        System.out.println("signature of Component has been extended");
    }
}

File: src/main/kotlin/Decorator.kt
interface IComponent {
    fun operate()
}

class ConcreteComponent : IComponent {
    override fun operate() = println("operating")
}

abstract class Decorator(private val component: IComponent) : IComponent by component

class ComponentDecorator(component: IComponent) : Decorator(component) {
    fun decoratedOperation() = println("signature of Component has been extended")
}

class AnotherDecoratedComponent(component: IComponent) : Decorator(component) {
    fun decoratedOperation() = println("signature of Component has been extended")
}

